jetty-server/src/main/java/org/eclipse/jetty/server/handler/AbstractHandlerContainer.java

```diff
@@ -135,4 +144,59 @@ public void setServer(Server server)
             for (Handler h : handlers)
                 h.setServer(server);
     }
+    
+
+    /* ------------------------------------------------------------ */
+    /** 
+     * Shutdown nested Gracefule handlers
+     * 
+     * @param futures A list of Futures which must also be waited on for the shutdown (or null)
+     * returns A MultiException to which any failures are added or null
+     */
+    protected void doShutdown(List<Future<Void>> futures) throws MultiException
+    {
+        MultiException mex = null;
+        
+        // tell the graceful handlers that we are shutting down
+        Handler[] gracefuls = getChildHandlersByClass(Graceful.class);
+        if (futures==null)
+            futures = new ArrayList<>(gracefuls.length);
+        for (Handler graceful : gracefuls)
+            futures.add(((Graceful)graceful).shutdown());
+
+        // Wait for all futures with a reducing time budget
+        long stopTimeout = getStopTimeout();
+        if (stopTimeout>0)
+        {
+            long stop_by=System.currentTimeMillis()+stopTimeout;
+            if (LOG.isDebugEnabled())
+                LOG.debug("Graceful shutdown {} by ",this,new Date(stop_by));
+
+            // Wait for shutdowns
+            for (Future<Void> future: futures)
+            {
+                try
+                {
+                    if (!future.isDone())
+                        future.get(Math.max(1L,stop_by-System.currentTimeMillis()),TimeUnit.MILLISECONDS);
+                }
+                catch (Exception e)
+                {
+                    if (!future.isDone() && future instanceof Callback)
```

Comment:
Calling `shutdown()` may return a `Callback`? If so, where is the callback succeeded?